doc/developer/design/20210412_postgres_sources.md

```diff
@@ -0,0 +1,210 @@
+# Postgres sources
+
+## Summary
+
+Many users are keen to use materialize with a dataset existing in a postgres
+upstream database. The current best-practice to ingest the data in realtime is
+to setup a Kafka cluster and use Debezium to convert the postgres CDC format
+into Kafka topics that can then be consumed by materialize.
+
+This design doc proposes to add a new type of source called `POSTGRES` where
+materialized directly connects to an upstream database and interprets the
+replication protocol, ingesting row updates in real time.
+
+Users should be able to try materialize out simply by running the materialize
+binary and pointing it to their postgres database, no extra infrastructure
+needed.
+
+## Goals
+
+The goal is to release a way for users to connect to a postgres database,
+covering most common setups and use cases. Since the configuration space is
+quite large this work focuses on a small set of features and makes a few
+assumptions that will get 80% of the way there.
+
+Specifically, the goals of postgres sources are to:
+
+* Connect to an upstream database with simple username/password authentication
+* Sync the initial state of the database and seamlessly switch to streaming
+* Guarantee that transaction boundaries across tables will be preserved
+* Support a suite of common column data types that cover most use cases
+* Materialize all synced tables by default. That is, all data must fit in RAM
```

Comment:
This is fine for MVP, but I'm concerned if this is a long-term constraint. For some databases, on-disk tables can be much larger than fits comfortably in memory.

Reply:
I 100% agree that supporting unmaterialized Pg sources is a crucial feature. We've discussed this on the team and I think there is consensus on that point.

Reply:
I agree. In order to support that we need to implement the dynamic resource sharing proposal that is quite a bit of work.

However I don't think this should block the release of the feature (maybe it should block version 1.0 specifically) because there are many usecases where the data fits comfortably in RAM.

Reply:
I think this means moving "unmaterialized sources" to a non-goal? At least, if we all agree that this is a temporary limitation rather than an intent of postgres sources.